CMP: add support for genm/genp messages with `id-it-caCerts`

DDvO opened the pull request:
This pertains to the CMP API, the CMP app, the doc, and the tests.
Doing so, found and fixed a bug in `gen_new()` in `cmp_msg.c` checking wrong ITAVs.

The CMP client may send a `genm` message with infoType `id-it-caCerts` (using option `-infotype caCerts`)
requesting CA certs, which the server provides in a `genp` message also with infoType `id-it-caCerts`.
The client performs sanity checks on the obtained certs and saves them in file given with `-cacertsout`.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

t8m commented on crypto/cmp/cmp_genm.c:
```diff
@@ -0,0 +1,183 @@
+    for (i = 0; i < sk_X509_num(certs /* may be NULL */); i++)
+        ret = ossl_X509_check(ctx, source,
+                              sk_X509_value(certs, i), type_CA, vpm)
+            && ret; /* Having 'ret' after the '&&', all certs are checked. */
```
Nit: IMO && should be aligned with ossl_...

DDvO replied:
Thanks for your comments!

`check-format.pl crypto/cmp/cmp_genm.c`, as well as the Emacs C mode, is happy with this indentation.
The idea is that on a multi-line assignment (and in similar other cases),
indentation is increased by the usual delta (+4), which is independent of the length of the variable name on the LHS.

paulidale replied:
I'd have indented this as @t8m suggests

DDvO replied:
I've reworked the assignment to generally make this code more readable.
As a by-product, the indentation issue disappears.

t8m commented on crypto/cmp/cmp_genm.c:
```diff
@@ -0,0 +1,183 @@
+    itavs = OSSL_CMP_exec_GENM_ses(ctx);
+    if (itavs == NULL) {
+        if (1 /*
+               * TOOD when PR #19205 is merged: replace '1' by
```
Typo: TOOD

DDvO replied:
Oops, fixed.

t8m commented on doc/man3/OSSL_CMP_exec_certreq.pod:
```diff
@@ -160,6 +169,8 @@ L<OSSL_CMP_MSG_http_perform(3)>
 
 The OpenSSL CMP support was added in OpenSSL 3.0.
 
+OSSL_CMP_get_caCerts() was added in OpenSSL 3.1.
```
OpenSSL 3.2 here

DDvO replied:
Updated.

t8m commented on include/openssl/cmp.h.in:
```diff
@@ -140,7 +140,6 @@ extern "C" {
 #  if OSSL_CMP_PKIFAILUREINFO_MAX_BIT_PATTERN > INT_MAX
 #   error CMP_PKIFAILUREINFO_MAX bit pattern does not fit in type int
 #  endif
-
```
Why this dropped line?

DDvO replied:
I deleted the empty line to make clear that the following decl
```
typedef ASN1_BIT_STRING OSSL_CMP_PKIFAILUREINFO;
```
is directly related.

t8m commented on include/openssl/cmp.h.in:
```diff
@@ -203,8 +202,8 @@ typedef ASN1_BIT_STRING OSSL_CMP_PKIFAILUREINFO;
 #  define OSSL_CMP_PKISTATUS_revocationWarning 4
 #  define OSSL_CMP_PKISTATUS_revocationNotification 5
 #  define OSSL_CMP_PKISTATUS_keyUpdateWarning 6
-
```
Why this?

DDvO replied:
Same reason as above.

DDvO commented:
@t8m ok now?

DDvO commented:
Asking @openssl/committers for 2nd review.

DDvO commented:
Rebased to fix (trivial) merge conflict in `libcrypto.num`.

Ping @openssl/committers for 2nd review.

DDvO commented:
Rebased again to fix trivial merge conflicts.

Asking again @openssl/committers for 2nd review.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

DDvO commented:
Can we please get this done now?
2nd approval is missing.

DDvO commented:
Fixed clashes of two `CMP_R` codes that came in when rebasing.

t8m reviewed (approved):
Still OK

DDvO commented:
@paulidale can you do the 2nd review/approval?

paulidale commented on apps/cmp.c:
```diff
@@ -2777,6 +2784,58 @@ static void print_status(void)
+            return res;
+        }
+        if (1 /*
+               * TOOD when PR #19205 is merged: replace '1' by
```
#19205 has been merged.

DDvO replied:
Good that you spotted this - resolved this TODO, as well as the two further ones referring to #19205.

paulidale commented on apps/cmp.c:
```diff
@@ -2777,6 +2784,58 @@ static void print_status(void)
+        if (!OSSL_CMP_get_caCerts(ctx, &cacerts))
+            return 0;
+
+        /* TODO possibly check authorization of sender/origin */
```
We try to avoid littering the code with TODOs...

DDvO replied:
Re-phrased.

paulidale commented on crypto/cmp/cmp_asn.c:
```diff
@@ -183,6 +184,37 @@ int OSSL_CMP_ITAV_push0_stack_item(STACK_OF(OSSL_CMP_ITAV) **itav_sk_p,
+        return 0;
+    }
+    *out = sk_X509_num(itav->infoValue.caCerts) > 0
+        ? itav->infoValue.caCerts : NULL;
```
indentation?

DDvO replied:
Well, indentation is actually correct here: +4 in multi-line assignments (like also in other similar cases).
I've just placed the `?` on the line before now, which should improve the appearance.

paulidale commented on crypto/cmp/cmp_genm.c:
```diff
@@ -0,0 +1,183 @@
+    req = NULL;
+    itavs = OSSL_CMP_exec_GENM_ses(ctx);
+    if (itavs == NULL) {
+        if (1 /*
```
This has been merged.

paulidale replied:
This is still here, either the code or comment is wrong.

paulidale reviewed (commented):
A fairly superficial review, I know nothing about what this is doing...

DDvO commented:
> A fairly superficial review, I know nothing about what this is doing...

Thank you @paulidale for anyway doing it!
I've just addressed all comments.

DDvO commented:
@paulidale can you approve now, given the above changes?

DDvO commented:
The two CI failures are unrelated.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

DDvO commented:
Rebased to fix a merge conflict.

DDvO commented:
Pinging again @paulidale - ok now?

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

paulidale commented:
@openssl/committers, looking for a second review some someone with half a clue here.

vdukhovni commented on apps/cmp.c:
```diff
@@ -2779,6 +2786,58 @@ static void print_status(void)
+            return 0;
+        }
+        return 1;
+    } else {
```
Is there really no suitable `opt_infotype` to check for in this branch?  An explicit choice feels safer than a default action.

vdukhovni commented on crypto/cmp/cmp_genm.c:
```diff
@@ -0,0 +1,183 @@
+    if (type_CA >= 0 && (ex_flags & EXFLAG_V1) == 0) {
+        int is_CA = (ex_flags & EXFLAG_CA) != 0;
+
+        if ((type_CA == 1) != is_CA) {
```
What is the expected result if the `type_CA > 1`? Would `!= 0` be better here?

vdukhovni commented on crypto/cmp/cmp_genm.c:
```diff
@@ -0,0 +1,183 @@
+    if (OSSL_CMP_CTX_get_status(ctx) != -1
+        /* TODO use OSSL_CMP_PKISTATUS_request when #19205 is merged */) {
+        ERR_raise_data(ERR_LIB_CMP, CMP_R_UNCLEAN_CTX,
+                       "should call CMPclient_reinit() before");
```
A less cryptic error message here?

vdukhovni commented on crypto/cmp/cmp_genm.c:
```diff
@@ -0,0 +1,183 @@
+
+    if ((n = sk_OSSL_CMP_ITAV_num(itavs)) <= 0) {
+        ERR_raise_data(ERR_LIB_CMP, CMP_R_INVALID_GENP,
+                       "infoType %s requested, but no ITAV included", desc);
```
It looks like `ITAV` is not a standard term in the context of CMP, but rather is OpenSSL-specific jargon.  While it may be reasonable to use in API types, ... where developers will have read the documentation, it should probably not leak into error messages seen by users.

If error messages from this protocol are to be presented to human users, they need a broad overhaul. They are generally only going to be understood by protocol experts, and need to be written for a broader audience.

DDvO replied:
Re-phrased to "response on genm requesting infoType %s does not include suitable value".

> If error messages from this protocol are to be presented to human users, they need a broad overhaul. They are generally only going to be understood by protocol experts, and need to be written for a broader audience.

Most OpenSSL lib error messages are rather cryptic and give pretty little detail such that they are typically hard to make use of, partly even for experts.
When I put error messages like in this PR, I try to be to-the-point and tend to give more detail, 
which hopefully will be help experts, while regular users anyway will not get much out of OpenSSL error queue dumps.

OTOH, the diagnostics of the CMP app should be digestible for regular users,
and with the improvement you suggested above I believe this has been achieved for this PR.

DDvO commented on apps/cmp.c:
```diff
@@ -2779,6 +2786,58 @@ static void print_status(void)
+        if (cacerts == NULL) {
+            CMP_warn("no CA certificate available");
+        } else if (save_free_certs(cacerts, opt_cacertsout, "CA") < 0) {
+            CMP_err1("Failed to store caCerts from genp in %s", opt_cacertsout);
```
Changed to "CA certificates".

DDvO commented on apps/cmp.c:
```diff
@@ -2779,6 +2786,58 @@ static void print_status(void)
+            CMP_warn1("No specific support for -infotype %s avaiable",
+                      opt_infotype_s);
+            if (req == NULL || !OSSL_CMP_CTX_push0_genm_ITAV(ctx, req)) {
+                CMP_err("Failed to create ITAV for genm");
```
Changed to `CMP_err1("Failed to create genm for -infotype %s", opt_infotype_s);`

DDvO commented on apps/cmp.c:
```diff
@@ -2779,6 +2786,58 @@ static void print_status(void)
+               * OSSL_CMP_CTX_get_status(ctx) != OSSL_CMP_PKISTATUS_request
+               */
+            )
+            CMP_err("Could not obtain valid response message on genm");
```
Changed to "Did not receive response on genm or genp is not valid".

DDvO commented:
Thank you @vdukhovni for your review comments.
I've meanwhile addressed all of them.

vdukhovni reviewed (commented):
I'll leave it to others to approve the PR, and gauge whether the updated error messages are now sufficiently user-friendly. I have no further issues to raise.

DDvO commented:
> @openssl/committers, looking for a second review from someone with half a clue here.

> I'll leave it to others to approve the PR, 
and gauge whether the updated error messages are now sufficiently user-friendly. 
I have no further issues to raise.

@paulidale can you approve now, since @vdukhovni is fine with the technical contents?

paulidale commented:
@vdukhovni has approval permission...

DDvO commented:
> @vdukhovni has approval permission...

@vdukhovni, would you mind approving yourself?
@paulidale appears fine with your remaining concerns, namely the user-friendliness of the (updated) error messages.

DDvO commented:
Since two weeks we are in a reviewer approval deadlock on this PR.
Can any of the @openssl/committers please help out.

t8m commented on apps/cmp.c:
```diff
@@ -2832,11 +2832,12 @@ static int do_genm(OSSL_CMP_CTX *ctx)
+        /* could check authorization of sender/origin at this point */
         if (cacerts == NULL) {
-            CMP_warn("no CA certificate available");
+            CMP_warn("no CA certificates providedd by server");
```
Typo: provided

DDvO replied:
oops, fixed

DDvO commented:
Rebased to fix merge conflicts

DDvO commented:
@openssl/committers, can we please get the 2nd approval done now?

DDvO commented:
Thanks a lot @paulidale for jumping in :tada:

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

DDvO commented:
Merged - thanks @t8m and @paulidale for your comments and approvals and @vdukhovni for your comments.
